### PR TITLE
validate sts roles on sts cluster upgrade

### PR DESCRIPTION
**Pull request by ciaranRoche**
Closes : https://issues.redhat.com/browse/SDA-4818

### Upgrade cluster mode manual 
```
 rosa upgrade cluster --cluster croche-upgrade --mode manual                                                                                              
? Version: 4.7.36
I: Ensuring cluster roles and policies are compatible with upgrade.
I: Preparing to upgrade account roles.
I: All policy files saved to the current directory
I: Run the following commands to upgrade the account role policies:

aws iam create-policy-version \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-ControlPlane-Role-Policy \
        --policy-document file://sts_instance_controlplane_permission_policy.json \
        --set-as-default

aws iam tag-policy \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-ControlPlane-Role-Policy

aws iam tag-role \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --role-name croche-upgrade-ControlPlane-Role

aws iam create-policy-version \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-Worker-Role-Policy \
        --policy-document file://sts_instance_worker_permission_policy.json \
        --set-as-default

aws iam tag-policy \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-Worker-Role-Policy

aws iam tag-role \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --role-name croche-upgrade-Worker-Role

aws iam create-policy-version \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-Support-Role-Policy \
        --policy-document file://sts_support_permission_policy.json \
        --set-as-default

aws iam tag-policy \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-Support-Role-Policy

aws iam tag-role \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --role-name croche-upgrade-Support-Role

aws iam create-policy-version \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-Installer-Role-Policy \
        --policy-document file://sts_installer_permission_policy.json \
        --set-as-default

aws iam tag-policy \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-Installer-Role-Policy

aws iam tag-role \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --role-name croche-upgrade-Installer-Role

aws iam create-policy-version \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-openshift-machine-api-aws-cloud-credentials \
        --policy-document file://openshift_machine_api_aws_cloud_credentials_policy.json \
        --set-as-default

aws iam tag-policy \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-openshift-machine-api-aws-cloud-credentials

aws iam create-policy-version \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-openshift-cloud-credential-operator-cloud-credent \
        --policy-document file://openshift_cloud_credential_operator_cloud_credential_operator_iam_ro_creds_policy.json \
        --set-as-default

aws iam tag-policy \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-openshift-cloud-credential-operator-cloud-credent

aws iam create-policy-version \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-openshift-image-registry-installer-cloud-credenti \
        --policy-document file://openshift_image_registry_installer_cloud_credentials_policy.json \
        --set-as-default

aws iam tag-policy \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-openshift-image-registry-installer-cloud-credenti

aws iam create-policy-version \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-openshift-ingress-operator-cloud-credentials \
        --policy-document file://openshift_ingress_operator_cloud_credentials_policy.json \
        --set-as-default

aws iam tag-policy \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-openshift-ingress-operator-cloud-credentials

aws iam create-policy-version \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-openshift-cluster-csi-drivers-ebs-cloud-credentia \
        --policy-document file://openshift_cluster_csi_drivers_ebs_cloud_credentials_policy.json \
        --set-as-default

aws iam tag-policy \
        --tags Key=rosa_openshift_version,Value=4.9 \
        --policy-arn arn:aws:iam::765374464689:policy/croche-upgrade-openshift-cluster-csi-drivers-ebs-cloud-credentia
I: Preparing to upgrade operator roles.
I: Account roles with prefix 'croche-upgrade' need to be upgraded before operator roles. Roles can be upgraded with the following command :

        rosa upgrade account-roles --prefix croche-upgrade

```

### Upgrade cluster mode auto
```
rosa upgrade cluster --cluster croche-upgrade --mode auto                                                                                                                                                                      1 ↵
? Version: 4.7.36
I: Ensuring cluster roles and policies are compatible with upgrade.
I: Preparing to upgrade account roles.
I: Starting to upgrade the policies
? Upgrade the 'croche-upgrade-Installer-Role' role polices to version 4.9? Yes
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-Installer-Role-Policy' to version '4.9'
? Upgrade the 'croche-upgrade-ControlPlane-Role' role polices to version 4.9? Yes
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-ControlPlane-Role-Policy' to version '4.9'
? Upgrade the 'croche-upgrade-Worker-Role' role polices to version 4.9? Yes
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-Worker-Role-Policy' to version '4.9'
? Upgrade the 'croche-upgrade-Support-Role' role polices to version 4.9? Yes
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-Support-Role-Policy' to version '4.9'
? Upgrade the operator role policies to version 4.9? Yes
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-openshift-cloud-credential-operator-cloud-credent' to version '4.9'
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-openshift-image-registry-installer-cloud-credenti' to version '4.9'
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-openshift-ingress-operator-cloud-credentials' to version '4.9'
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-openshift-cluster-csi-drivers-ebs-cloud-credentia' to version '4.9'
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-openshift-machine-api-aws-cloud-credentials' to version '4.9'
I: Preparing to upgrade operator roles.
I: Starting to upgrade the policies
? Upgrade the 'croche-upgrade-u3b5-openshift-image-registry-installer-cloud-cre' operator role policy to version 4.9? Yes
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-openshift-image-registry-installer-cloud-credenti' to version '4.9'
? Upgrade the 'croche-upgrade-u3b5-openshift-ingress-operator-cloud-credentials' operator role policy to version 4.9? Yes
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-openshift-ingress-operator-cloud-credentials' to version '4.9'
? Upgrade the 'croche-upgrade-u3b5-openshift-cluster-csi-drivers-ebs-cloud-cred' operator role policy to version 4.9? Yes
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-openshift-cluster-csi-drivers-ebs-cloud-credentia' to version '4.9'
? Upgrade the 'croche-upgrade-u3b5-openshift-machine-api-aws-cloud-credentials' operator role policy to version 4.9? Yes
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-openshift-machine-api-aws-cloud-credentials' to version '4.9'
? Upgrade the 'croche-upgrade-u3b5-openshift-cloud-credential-operator-cloud-cr' operator role policy to version 4.9? Yes
I: Upgraded policy with ARN 'arn:aws:iam::765374464689:policy/croche-upgrade-openshift-cloud-credential-operator-cloud-credent' to version '4.9'
? Please input desired date in format yyyy-mm-dd: 2021-11-05
? Please input desired UTC time in format HH:mm: 13:50
? Node draining: 1 hour
I: Upgrade successfully scheduled for cluster 'croche-upgrade'

```

### Upgrade cluster no mode 
```
rosa upgrade cluster --cluster croche-upgrade 
? Version: 4.7.36
I: Ensuring cluster roles and policies are compatible with upgrade.
I: Cluster Roles are not valid with upgrade version 4.7.36. Run the following command(s) to upgrade Cluster Roles:

        rosa upgrade account-roles --prefix croche-upgrade
        rosa upgrade operator-roles --cluster croche-upgrade

```

### CODE REVIEW

File: pkg/aws/policies.go [hunk 1]
```diff
@@ -448,6 +493,102 @@ func (c *awsClient) HasCompatibleVersionTags(iamTags []*iam.Tag, version string)
...
+	}
+	return true, nil
+}
+
+// ValidateClusterOperatorRolesVersion ensures all operator roles for a cluster is compatible with a specific
+// openshift version
+func (c *awsClient) ValidateClusterOperatorRolesVersion(cluster *cmv1.Cluster, version string, creator *Creator) (
+	bool, error) {
+	operatorRoles, hasOperatorRoles := cluster.AWS().STS().GetOperatorIAMRoles()
+	if !hasOperatorRoles {
```
Comment (pvasant) [inline pkg/aws/policies.go:575]: This scenario is not possible now

File: pkg/aws/policies.go [hunk 2]
```diff
@@ -448,6 +493,102 @@ func (c *awsClient) HasCompatibleVersionTags(iamTags []*iam.Tag, version string)
...
+		return false, errors.Errorf("Can not validate cluster roles, operator roles for cluster '%s' does not exist",
+			cluster.Name())
+	}
+	for _, role := range operatorRoles {
+		isCompatible, err := c.validateRoleUpgradeVersionCompatibility(getRoleNameFromARNSplit(role.RoleARN()),
```
Comment (pvasant) [inline pkg/aws/policies.go:580]: I mentioned in other PR as well we cannot use the same method for operator role and account role

File: cmd/create/accountroles/cmd.go [hunk 3]
```diff
@@ -275,7 +265,7 @@ func run(cmd *cobra.Command, argv []string) {
 		commands := buildCommands(prefix, permissionsBoundary, creator.AccountID)
 		fmt.Println(commands)
 	default:
```
Comment (vkareh) [inline cmd/create/accountroles/cmd.go:267]: The previous `case` blocks here should use `ocm.ModeManual`, etc.

File: pkg/ocm/flag.go [hunk 4]
```diff
@@ -40,6 +49,19 @@ func AddClusterFlag(cmd *cobra.Command) {
...
 
+func AddModeFlag(cmd *cobra.Command) {
+	cmd.Flags().StringVarP(
+		&mode,
+		"mode",
+		"m",
+		"",
+		"How to perform the operation. Valid options are:\n"+
+			"auto: To automatically apply command actions\n"+
+			"manual: To output commands which can be used to manually complete command actions",
```
Comment (vkareh) [inline pkg/ocm/flag.go:60]: I would change this to:
- auto: Resource changes will be automatic applied using the current AWS account
- manual: Commands necessary to modify AWS resources will be output to be run manually

File: pkg/ocm/flag.go [hunk 5]
```diff
@@ -21,12 +21,21 @@ import (
...
 
 var clusterKey string
+var mode string
+
+const (
+	ModeAuto   = "auto"
+	ModeManual = "manual"
+)
+
+var Modes = []string{ModeAuto, ModeManual}
```
Comment (vkareh) [inline pkg/ocm/flag.go:38]: I'd say this is probably not a good candidate for an "ocm" flag, more like an "aws" flag, but that's me /nitpicking :)

**Comment by vkareh** [comment]
@ciaranRoche - this is fantastic, thanks! I will test this for functionality, but the code looks good to me. I only have a few nitpicks posted above

File: cmd/upgrade/cluster/cmd.go [hunk 6]
```diff
@@ -215,6 +225,59 @@ func run(cmd *cobra.Command, _ []string) {
 		interactive.Enable()
 	}
 
+	// if cluster is sts validate roles are compatible with upgrade version
+	_, isSTS := cluster.AWS().STS().GetRoleARN()
```
Comment (vkareh) [inline cmd/upgrade/cluster/cmd.go:229]: Can you please add a validation like:
```
if !isSTS && mode != "" {
  reporter.Errorf("The 'mode' option is only supported for STS clusters")
  os.Exit(1)
}
```
But add it really early (before _any_ interactive prompt is shown)

**Pull request merged**